src/main.cpp

```diff
@@ -4754,24 +4747,26 @@ bool ProcessBlock(CNode* pfrom, CBlock* pblock, bool generated_by_me)
             ResetTimerMain("CheckForFutileSync");
         }
 
-
-        CBlock* pblock2 = new CBlock(*pblock);
-        if (WalletOutOfSyncByMoreThan2000Blocks() || fTestNet)
```

Comment:
TL;DR: This check is removed and the allocation of the block doesn't happen until it's inserted into the map.